spec/src/main/asciidoc/converters.asciidoc

```diff
@@ -92,3 +92,50 @@ If no built-in nor custom `Converter` for a requested Type `T`, an implicit Conv
 * The target type {@code T} has a public Constructor with a String parameter, or
 * The target type {@code T} has a {@code public static T parse(CharSequence)} method
 
+=== Empty value
+
+A config property can be explicitly configured with an empty value (the empty String `""`).  If a property
+is missing and no default value exists for that property, it is considered to be empty.
+
+The `Config.getValue()` method will normally throw a `NoSuchElementException` when a configuration value
+is empty.  In order to allow empty values to be returned, either the `Config.getOptionalValue()`
+method must be used, or a custom converter must be specified which provides a specific value for
+the case where an empty value is returned (either implicitly due to the property being missing, or
+explicitly due to an empty value being specified for a configuration property).
+
+[NOTE]
+This behaviour allows users to _unconfigure_ a property. If a property is configured in a lower-ordinal ConfigSource with
+an undesirable value, you can set the property value to an empty string in a higher-ordinal ConfigSource to discard the lower-ordinal value.
+
+When processing array or collection values, any empty array or collection segments must be excluded from the
+resultant array or collection.
+
+[[empty_value_table]]
+.Examples of Config API behavior for empty values
+[options="header"]
+|=======================
+| Property value | Output type | `Config` method | Output result
+| missing     | `String` | `getValue` | throws `NoSuchElementException`
+| `""` (empty string) | `String` | `getValue` | throws `NoSuchElementException`
+| `" "`       | `String` | `getValue` | `" "`
+| `"foo"`     | `String`   | `getValue` |  `"foo"`
+| missing     | `String[]` | `getValue` | throws `NoSuchElementException`
+| `""`        | `String[]` | `getValue` | throws `NoSuchElementException`
+| `" "`       | `String[]` | `getValue` | `{ " " }`
+| `","`       | `String[]` | `getValue` | throws `NoSuchElementException`
+| `",,"`      | `String[]` | `getValue` | throws `NoSuchElementException`
+| `"foo"`     | `String[]` | `getValue` | `{ "foo" }`
+| `"foo,bar"` | `String[]` | `getValue` | `{ "foo", "bar" }`
+| `"foo,"`    | `String[]` | `getValue` | `{ "foo" }`
+| `",bar"`    | `String[]` | `getValue` | `{ "bar" }`
+| `",bar,"`   | `String[]` | `getValue` | `{ "bar" }`
+| missing     | `String` |`getOptionalValue` | `Optional.empty()`
+| `""`        | `String` | `getOptionalValue` | `Optional.empty()`
+| `" "`        | `String` | `getOptionalValue` | `Optional.of(" ")`
+| `"foo"`     | `String` | `getOptionalValue` | `Optional.of("foo")`
+| missing | `String[]` | `getOptionalValue` | `Optional.empty()`
+| `""` | `String[]` | `getOptionalValue` | `Optional.empty()`
```

Comment:
This section of documentation seems contradictory.  The section begins by stating "A config property can be explicitly configured with an empty value (the empty String `""`)", which is something I'm in favor of.  But it later goes on to state that the returned Optional then ends up without a value: `Optional.empty()`.  Per the documentation of java.util.Optional, this clearly indicates missing, rather than a value that is an empty String/String[]/whatever.  This makes usage of the Optional interface counter-intuitive.  Consider what will happen when MicroProfile users write code per the intended meaning of java.util.Optional that is defined within java.util.Optional's own JavaDoc. When users write,
```
   optional.ifPresent(value -> consumeValue(value));
```
they expect the Consumer to run with the value that was configured: empty value or otherwise.  However, that won't happen if MP Config halfway repurposes Optional.empty, in which case it would behave as if there were no value present at all - not an empty value.

Similarly, when users write,
```
   propValue = optional.orElse(new String[] { DEFAULT_ITEM1, DEFAULT_ITEM2 });
```
they expect to actually get an empty String array value assigned if their user configured to have an empty string array.  However, if MP Config repurposes `Optional.empty()` as proposed here then they'll end up with the supplied parameter instead.

In the JavaDOC API for java.util.Optional, `Optional.empty()` very clearly indicates missing, both in how it is documented (stating that ["No value is present for this Optional"](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#empty--)), as well as in the behavior of the various operations that are available on java.util.Optional.  I strongly caution against altering its meaning when used within MP Config.  Doing so would stray from all other expected usage and leave MicroProfile with a confusing and difficult-to-use spec that is inconsistent with Java and will be the source of numerous bugs.

This basically gets at the underlying problem that I've had trouble with all along in the earlier discussions.  Missing values and values for instances that are empty are fundamentally two distinct things and cannot share a single representation without causing inconsistency and a source of confusion/error.   An Optional that is empty (has no value) is different than an Optional that has a value which is an empty instance of something: the empty string, the empty set, an empty array or list, ...

Reply:
> This section of documentation seems contradictory. The section begins by stating "A config property can be explicitly configured with an empty value (the empty String `""`)", which is something I'm in favor of. But it later goes on to state that the returned Optional then ends up without a value: `Optional.empty()`. Per the documentation of java.util.Optional, this clearly indicates missing, rather than a value that is an empty String/String[]/whatever.

The method is literally called `empty()`, which to me pretty clearly means "empty".  So we might have to agree to disagree on this one.

> This makes usage of the Optional interface counter-intuitive. Consider what will happen when MicroProfile users write code per the intended meaning of java.util.Optional that is defined within java.util.Optional's own JavaDoc. When users write,
> 
> ```
>    optional.ifPresent(value -> consumeValue(value));
> ```
> 
> they expect the Consumer to run with the value that was configured: empty value or otherwise.

I don't know that this would be intuitive.  Consider for example the case where `value` is not a `String`, but rather, say, a `Double` or a `LocalDate` or something.

> Similarly, when users write,
> 
> ```
>    propValue = optional.orElse(new String[] { DEFAULT_ITEM1, DEFAULT_ITEM2 });
> ```
> 
> they expect to actually get an empty String array value assigned if their user configured to have an empty string array. However, if MP Config repurposes `Optional.empty()` as proposed here then they'll end up with the supplied parameter instead.

That's what default value is for.  I don't think it's that confusing to say "if you have a default value, use the default value method".

> In the JavaDOC API for java.util.Optional, `Optional.empty()` very clearly indicates missing, both in how it is documented (stating that ["No value is present for this Optional"](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#empty--)), as well as in the behavior of the various operations that are available on java.util.Optional. I strongly caution against altering its meaning when used within MP Config.

We will now define an empty input string as being "explicitly missing", which is completely within the documented definition of the method.  Again you've latched on to the behavior of `String`, which has its own natural empty representation.  Most types do not.  If I give a property value like `my.int.value=`, I am clearly and unambiguously saying "this value is now explicitly cleared by my request".  Whereas by specifying a default value, I am clearly saying "if no value is given for this property, it has this value", where the default value will generally be non-empty.  So in this case it is not missing at all.

Once we look at property expansion, which is likely to immediately follow this issue, the traditional idea of "missing" becomes even more nonsensical.  When a property is expanded, it may reference other properties which in turn may be expanded, and so on.  These properties might not be present in the traditional sense, but they may be associated with default values.  The only way to accomplish this is by having a bottom-priority config source which contains all the default values for properties which have them.  Once you do this however, then the idea of detecting "missing" by `Optional` no longer works, no matter how intuitive or unintuitive it may appear to you.  It simply ceases to be an option, because any "missing" value will appear to be "present" by the `Optional` API due to the default value resolution.  The proposed specification change, however, integrates flawlessly with this case.

In any case this is the only viable specification option that has been found or that seems possible for fixing the behavioral inconsistencies without introducing other inconsistencies or blocking the way to property expansion.  An alternative proposal would be welcome, if it satisfies all the same cases.  If you open such a PR, reference this issue.  However, I simply do not believe it is possible to solve the underlying issues without a common representation of empty values, and distinct means for handling empty and missing.  And doing this will, as far as I can tell, inevitably require that empty and missing-with-no-default must necessarily be represented by the same value; in other words, you'll end up in the same place once you have accounted for all the hard constraints.

> Doing so would stray from all other expected usage and leave MicroProfile with a confusing and difficult-to-use spec that is inconsistent with Java and will be the source of numerous bugs.

That is just pure hyperbole, which I completely disagree with, so I'm going to leave that aside.

> This basically gets at the underlying problem that I've had trouble with all along in the earlier discussions. Missing values and values for instances that are empty are fundamentally two distinct things and cannot share a single representation without causing inconsistency and a source of confusion/error. An Optional that is empty (has no value) is different than an Optional that has a value which is an empty instance of something: the empty string, the empty set, an empty array or list, ...

Right, this is the essential problem of having multiple empty representations: every converter ends up with different behavior.  In addition, of course, some types have no empty representation, so sometimes empty values will fail and sometimes they will not, making it impossible to use an empty input string as a way to unset a property, which is the only intuitive solution to that problem.  So, we unify on exactly one representation of empty.  The examples should be fairly easy to follow, so I really don't anticipate much justifiable confusion.

I'm sorry that this specification change does not meet your liking.  However, it cannot be denied that the change does work with all of the existing and suggested use cases as well as making it possible to support property expansion.  In this case I think it is important to design for correctness first.  If there is some possible change that might improve usability in your eyes, without sacrificing correctness, then I'm eager to hear about it.  But otherwise, we have to accept this change and move forward.

As such I will not accept or respond to any more subjective or hyperbolic assessments of the change.  Any objective ideas for usability enhancements can be proposed as pull requests to one or the other of the branches on these pull requests (https://github.com/dmlloyd/microprofile-config/tree/empty-values-option-1 or https://github.com/dmlloyd/microprofile-config/tree/empty-values-option-2) and they will be considered fairly as long as they do not cause a regression that invalidates any of the preestablished use cases or behaviors.

Reply:
> The method is literally called `empty()`, which to me pretty clearly means "empty". So we might have to agree to disagree on this one.

Neither of us gets to decide for ourselves what [Optional.empty()](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#empty--) means.  This isn't a matter of opinion that we can agree to disagree on.  java.util.Optional is a standard, it is JavaDoc'ed, and it is already implemented with a behavior that won't change based on any arguments we make in this pull.

> If there is some possible change that might improve usability in your eyes, without sacrificing correctness, then I'm eager to hear about it. But otherwise, we have to accept this change and move forward.
> 
> As such I will not accept or respond to any more subjective or hyperbolic assessments of the change. Any objective ideas for usability enhancements can be proposed as pull requests to one or the other of the branches on these pull requests

I realize, as someone who lacks any experience in the implementation of MicroProfile Config, that my comments are only suggestions.  It sounds like acceptance of this pull, or the other one, with the same identical section, is already a foregone conclusion.  To that end, the best I can do, having conceded that the MP config spec will be changing behavior, is to help you align the language in the documentation with what these changes are actually doing.  I would rather see the MicroProfile Config spec be meaningful and well understood in the behavior that you actually get from it, rather than for it to portray that it is giving me a behavior which I requested but am not actually getting.

To accomplish that, please rewrite this section, previously titled "Empty value", to instead be:

```
=== Missing value

A config property can be explicitly configured to indicate that a property is missing by specifying the empty String (`""`).  The property
is considered to be missing if no default value exists for that property.

The `Config.getValue()` method will normally throw a `NoSuchElementException` when a configuration value
is missing.  Empty values (such as the empty string `""` or a zero-length array) cannot be returned
unless a custom converter is specified which provides a specific value for
the case where an empty value is desired.

[NOTE]
This behaviour allows users to _unconfigure_ a property. If a property is configured in a lower-ordinal ConfigSource with
an undesirable value, you can set the property value to an empty string, which means missing, in a higher-ordinal ConfigSource to discard the lower-ordinal value.

...
```

The remaining portion of this section can then remain as is.

The MicroProfile spec that I made the original request for empty array on behalf of, will then need to work around this change in behavior by defining its own constant for users to specify when they want empty array.

Reply:
> > The method is literally called `empty()`, which to me pretty clearly means "empty". So we might have to agree to disagree on this one.
> 
> Neither of us gets to decide for ourselves what [Optional.empty()](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#empty--) means. This isn't a matter of opinion that we can agree to disagree on. java.util.Optional is a standard, it is JavaDoc'ed, and it is already implemented with a behavior that won't change based on any arguments we make in this pull.

Well, surely we agree that the word "empty" means "empty", regardless of subjective context.  There otherwise is no objective relationship that I can discern between `Optional`'s documented behavior and the use to which it is put here: any use of the word "empty" in the `Optional` documentation seems to refer to the `Optional` itself, which is defined only as "A container object which may or may not contain a non-`null` value".  In fact the word "missing" never actually seems to appear in the `Optional` JavaDoc, so I'm not quite sure what you're referring to here.

> > If there is some possible change that might improve usability in your eyes, without sacrificing correctness, then I'm eager to hear about it. But otherwise, we have to accept this change and move forward.
> > As such I will not accept or respond to any more subjective or hyperbolic assessments of the change. Any objective ideas for usability enhancements can be proposed as pull requests to one or the other of the branches on these pull requests
> 
> I realize, as someone who lacks any experience in the implementation of MicroProfile Config, that my comments are only suggestions. It sounds like acceptance of this pull, or the other one, with the same identical section, is already a foregone conclusion. To that end, the best I can do, having conceded that the MP config spec will be changing behavior, is to help you align the language in the documentation with what these changes are actually doing. I would rather see the MicroProfile Config spec be meaningful and well understood in the behavior that you actually get from it, rather than for it to portray that it is giving me a behavior which I requested but am not actually getting.
> 
> To accomplish that, please rewrite this section, previously titled "Empty value", to instead be:

Sure, this change seems quite reasonable.  I'll make the changes.

> The MicroProfile spec that I made the original request for empty array on behalf of, will then need to work around this change in behavior by defining its own constant for users to specify when they want empty array.

That is not true.  If you have a look at the description of #446, you should be able to extrapolate how to accomplish this use case (n.b. I'm understanding the use case to be "have a non-empty default value with a special meaning which might not be otherwise representable as a string value") without requiring a special constant for users to specify.  The point of the proposed solution is to be able to meet use cases like this one without requiring the user to go through any unusual acrobatics.

Of course you're also free to have a special constant, if you feel that gives a better user experience for that particular case.  In particular it allows the user to override a lower priority value with a value representing "whatever the default is", which is not otherwise possible.  But it is definitely not required otherwise.

Note that having a representable default value - even if the literal string isn't known until the program runs - will allow the value to participate in property expansion, which otherwise it would not be able to do.

Reply:
I've made the changes, and I marked you (@njr-11) as co-author using the email address you've used on other MicroProfile commits.  Please let me know if this is OK.

Reply:
> That is not true. If you have a look at the description of #446, you should be able to extrapolate how to accomplish this use case (n.b. I'm understanding the use case to be "have a non-empty default value with a special meaning which might not be otherwise representable as a string value") without requiring a special constant for users to specify. The point of the proposed solution is to be able to meet use cases like this one without requiring the user to go through any unusual acrobatics.

Here is what I see in #446 for this scenario,

> // approach 1 (using implementation option 1)
String value = config.getOptionalValueWithDefault("prop6.not.specified", String.class, "default-value").orElse("");
// approach 2 (using implementation option 2)
String value = config.getAccessor("prop6.not.specified")
        .convertedForType(String.class)
        .withDefault("default-value")
        .optional()
        .getValue()
        .orElse("");

In the case of approach 1, the getOptionalValueWithDefault method can never return Optional.empty() because "default-value" is not null.  (See the JavaDoc for getOptionalValueWithDefault).  That means `orElse` is a no-op here.

In the case of approach 2, the Optional that is returned by `ConfigAccessor<Optional<String>>.getValue()` similarly either has a value of "default-value" or the configured value (absent any default) and so the `orElse` is again a no-op.

To get this working properly, getOptionalValueWithDefault [1] and getValue [2] would need to behave differently and instead return Optional.empty(), ignoring the non-null default, but that would bust the case where the user specified "" to unconfigure a lower-order config source and let the value default.

This goes back to the problem that "" cannot mean 2 different things at once.  With MP Config now claiming "" for unconfiguring/missing value, it would simply be wrong for another MicroProfile spec to claim that it must mean something else.  That's why the new constant will be needed.

Reply:
> > That is not true. If you have a look at the description of #446, you should be able to extrapolate how to accomplish this use case (n.b. I'm understanding the use case to be "have a non-empty default value with a special meaning which might not be otherwise representable as a string value") without requiring a special constant for users to specify. The point of the proposed solution is to be able to meet use cases like this one without requiring the user to go through any unusual acrobatics.
> 
> Here is what I see in #446 for this scenario,
> 
> > // approach 1 (using implementation option 1)
> > String value = config.getOptionalValueWithDefault("prop6.not.specified", String.class, "default-value").orElse("");
> > // approach 2 (using implementation option 2)
> > String value = config.getAccessor("prop6.not.specified")
> > .convertedForType(String.class)
> > .withDefault("default-value")
> > .optional()
> > .getValue()
> > .orElse("");
> 
> In the case of approach 1, the getOptionalValueWithDefault method can never return Optional.empty() because "default-value" is not null. (See the JavaDoc for getOptionalValueWithDefault). That means `orElse` is a no-op here.

That's not correct.  The default value is returned if no configuration source specifies a that configuration key.  The empty optional is returned if the configuration source with the highest priority for a given key explicitly specifies an empty value for that key (e.g. `keyname=`).  That's what this sentence means: "The return value will be empty if the value is empty or if the property is not found and the default value given is `null`".  If the default value is not null, then the other behavior applies: "...using the given default value if the property is not found".  Specifying an empty value in this case does not cause the value to be "not found", rather it is found to have an empty value, rather than whatever the default value is.

It all works, honest!

> In the case of approach 2, the Optional that is returned by `ConfigAccessor<Optional<String>>.getValue()` similarly either has a value of "default-value" or the configured value (absent any default) and so the `orElse` is again a no-op.

This is the same as above: if the key is not specified, the default value is used; otherwise, if the value is empty, the empty optional is returned.

Does that make sense?

Reply:
> Specifying an empty value in this case does not cause the value to be "not found", rather it is found to have an empty value, rather than whatever the default value is.

Ok - I was missing that detail about "found to have an empty value" vs "not found".  I still have trouble grasping how it can on the one hand be found to have a value, but on the other hand, that value isn't the value that is considered present within the Optional.  Are you certain users will be able to understand the minute details needed to properly use these methods?

Reply:
I would have thought so.  But, any ideas as to how to further clarify the language would be welcome.  I'll spend some time today trying to improve the verbiage if I can.